src/algebra/free.lean

```diff
@@ -2,173 +2,323 @@
-@[simp] lemma traverse_eq (x) : free_magma.traverse F x = traverse F x := rfl
+@[simp, to_additive]
+lemma traverse_pure (x) : traverse F (pure x : free_magma α) = pure <$> F x := rfl
+
+@[simp, to_additive]
+lemma traverse_pure' : traverse F ∘ pure = λ x, (pure <$> F x : m (free_magma β)) := rfl
+
+@[simp, to_additive]
+lemma traverse_mul (x y : free_magma α) : traverse F (x * y) = (*) <$> traverse F x <*> traverse F y := rfl
+
+@[simp, to_additive]
+lemma traverse_mul' : function.comp (traverse F) ∘ @has_mul.mul (free_magma α) _ = λ x y, (*) <$> traverse F x <*> traverse F y := rfl
+
+@[simp, to_additive]
+lemma traverse_eq (x) : free_magma.traverse F x = traverse F x := rfl
 
-@[simp] lemma mul_map_seq (x y : free_magma α) : ((*) <$> x <*> y : id (free_magma α)) = (x * y : free_magma α) := rfl
+@[simp, to_additive]
+lemma mul_map_seq (x y : free_magma α) : ((*) <$> x <*> y : id (free_magma α)) = (x * y : free_magma α) := rfl
 
+@[to_additive]
 instance : is_lawful_traversable free_magma.{u} :=
-{ id_traverse := λ α x, free_magma.induction_on x (λ x, rfl)
+{ id_traverse := λ α x, free_magma.rec_on'' x (λ x, rfl)
     (λ x y ih1 ih2, by rw [traverse_mul, ih1, ih2, mul_map_seq]),
-  comp_traverse := λ F G hf1 hg1 hf2 hg2 α β γ f g x, free_magma.induction_on' x
+  comp_traverse := λ F G hf1 hg1 hf2 hg2 α β γ f g x, free_magma.rec_on'' x
     (λ x, by resetI; simp only [traverse_pure, traverse_pure'] with functor_norm)
     (λ x y ih1 ih2, by resetI; rw [traverse_mul, ih1, ih2, traverse_mul];
       simp only [traverse_mul'] with functor_norm),
-  naturality := λ F G hf1 hg1 hf2 hg2 η α β f x, free_magma.induction_on' x
+  naturality := λ F G hf1 hg1 hf2 hg2 η α β f x, free_magma.rec_on'' x
     (λ x, by simp only [traverse_pure] with functor_norm)
     (λ x y ih1 ih2, by simp only [traverse_mul] with functor_norm; rw [ih1, ih2]),
-  traverse_eq_map_id := λ α β f x, free_magma.induction_on x (λ _, rfl)
+  traverse_eq_map_id := λ α β f x, free_magma.rec_on'' x (λ _, rfl)
     (λ x y ih1 ih2, by rw [traverse_mul, ih1, ih2, map_mul', mul_map_seq]; refl) }
 
 end category
 
-instance [decidable_eq α] : decidable_eq (free_magma α)
-| (of p)    (of x)    := decidable_of_iff (p = x) ⟨congr_arg of, of.inj⟩
-| (of p)    (mul x y) := is_false $ λ H, free_magma.no_confusion H
-| (mul p q) (of x)    := is_false $ λ H, free_magma.no_confusion H
-| (mul p q) (mul x y) := @decidable_of_iff (mul p q = mul x y) (p = x ∧ q = y) ⟨λ ⟨hpx, hqy⟩, hpx ▸ hqy ▸ rfl, mul.inj⟩
-    (@and.decidable _ _ (decidable_eq p x) (decidable_eq q y))
-
-def repr' [has_repr α] : free_magma α → string
-| (of x) := repr x
-| (mul x y) := "( " ++ repr' x ++ " * " ++ repr' y ++ " )"
-
-instance [has_repr α] : has_repr (free_magma α) := ⟨repr'⟩
-
-def length : free_magma α → ℕ
-| (of x)    := 1
-| (mul x y) := length x + length y
-
 end free_magma
 
+instance {α : Type u} [decidable_eq α] : decidable_eq (free_magma α)
```

Comment:
Does `@[derive decidable_eq]` on the definition of `free_magma` not work?

Reply:
Or does this have to be done manually for `to_additive` for some reason?

Reply:
Done.